modules/renter/download.go

```diff
@@ -1,182 +1,552 @@
 	"errors"
-	"fmt"
-	"io"
 	"os"
-	"strings"
-	"sync/atomic"
+	"sync"
 	"time"
 
+	"github.com/NebulousLabs/Sia/build"
 	"github.com/NebulousLabs/Sia/crypto"
 	"github.com/NebulousLabs/Sia/modules"
-	"github.com/NebulousLabs/Sia/modules/renter/contractor"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+const (
+	defaultFilePerm = 0666
 )
 
 var (
+	errPrevErr            = errors.New("download could not be completed due to a previous error")
 	errInsufficientHosts  = errors.New("insufficient hosts to recover file")
 	errInsufficientPieces = errors.New("couldn't fetch enough pieces to recover data")
+
+	// maxActiveDownloadPieces determines the maximum number of pieces that are
+	// allowed to be concurrently downloading. More pieces means more
+	// parallelism, but also more RAM usage.
+	//
+	// TODO: Allow this number to be established in the renter settings.
+	maxActiveDownloadPieces = build.Select(build.Var{
+		Standard: int(25),
+		Dev:      int(10),
+		Testing:  int(5),
+	}).(int)
 )
 
-// A fetcher fetches pieces from a host. This interface exists to facilitate
-// easy testing.
-type fetcher interface {
-	// pieces returns the set of pieces corresponding to a given chunk.
-	pieces(chunk uint64) []pieceData
+type (
+	// chunkDownload tracks the progress of a chunk. The chunk download object
+	// should only be read or modified inside of the main download loop thread.
+	chunkDownload struct {
+		download *download
+		index    uint64 // index of the chunk within the download
 
-	// fetch returns the data specified by piece metadata.
-	fetch(pieceData) ([]byte, error)
-}
+		// completedPieces contains information about which pieces have been
+		// successfully downloaded.
+		//
+		// workerAttempts contains a list of workers that are able to fetch a
+		// piece of the chunk, mapped to an indication of whether or not they
+		// have tried to fetch a piece of the chunk.
+		completedPieces map[uint64][]byte
+		workerAttempts  map[types.FileContractID]bool
+	}
+
+	// A download is a file download that has been queued by the renter.
+	download struct {
+		// Progress variables.
+		dataReceived     uint64
+		downloadComplete bool
+		downloadErr      error
+		finishedChunks   []bool
+
+		// Timestamp information.
+		completeTime time.Time
+		startTime    time.Time
+
+		// Static information about the file - can be read without a lock.
+		chunkSize   uint64
+		destination string
+		erasureCode modules.ErasureCoder
+		fileSize    uint64
+		masterKey   crypto.TwofishKey
+		numChunks   uint64
+		pieceSet    []map[types.FileContractID]pieceData
+		siapath     string
+
+		// Syncrhonization tools.
+		downloadFinished chan error
+		mu               sync.Mutex
+	}
+
+	// downloadState tracks all of the stateful information within the download
+	// loop, primarily used to simplify the use of helper functions. There is
+	// no thread safety with the download state, as it is only ever accessed by
+	// the primary download loop thread.
+	downloadState struct {
```

Comment:
logging is the biggest thing I think. But overall the hierarchy is pretty thoroughly overlapped. It would take a pretty big re-write to fix that though.